filetype: TI linear assembly files are not recognized

Freed-Wu opened the pull request:
Problem: some TI DSP related files are not recognized
Solution: recognize some TI DSP related files as their filetypes

chrisbra commented on runtime/filetype.vim:
```diff
@@ -233,9 +233,9 @@ au BufNewFile,BufRead *.fb			setf freebasic
+" *.cmd is close to a Batch file, but on OS/2 Rexx files and TI DSP linker command files also use *.cmd.
 au BufNewFile,BufRead *.cmd
-	\ if getline(1) =~ '^/\*' | setf rexx | else | setf dosbatch | endif
+	\ if getline(1) =~ '/\*\s\+\S\+\.cmd' | setf cmd | elseif getline(1) =~ '^/\*' | setf rexx | else | setf dosbatch | endif
```
Your example file has the script name on line 2. Also I find this is a strange condition to detect TI DSP file. I can imagine that batch-scripts may also have similar comments which may match `*.cmd`. Also I find cmd as filetype a bit too generic. How does this related to "TI Linker"?

Freed-Wu replied:
dosbatch will not use `/* XXX.cmd` because `/* XXX */` is comment of tild.

chrisbra commented on runtime/syntax/cmd.vim:
```diff
@@ -0,0 +1,11 @@
+endif
+
+" TODO: support C macro defines
+setlocal syntax=ld
```
use `runtime! syntax/ld.vim`

chrisbra commented on runtime/syntax/gel.vim:
```diff
@@ -0,0 +1,11 @@
+endif
+
+" TODO: support gel specific keywords
+setlocal syntax=c
```
same here

chrisbra commented on runtime/syntax/gel.vim:
```diff
@@ -0,0 +1,11 @@
+" TODO: support gel specific keywords
+setlocal syntax=c
+
+let b:current_syntax = "cmd"
```
is that intentionally set to cmd?

chrisbra commented on runtime/syntax/sa.vim:
```diff
@@ -0,0 +1,11 @@
+" TODO: not same asm, just better than none
+setlocal syntax=asm
+
+let b:current_syntax = "cmd"
```
is that intentionally set to cmd?

chrisbra commented on runtime/syntax/sa.vim:
```diff
@@ -0,0 +1,11 @@
+endif
+
+" TODO: not same asm, just better than none
+setlocal syntax=asm
```
`runtime! syntax/asm.vim`

chrisbra reviewed (commented):
I find those filetypes all a bit arbitrary. Can't you use `tigel`, `tiasm`, `tild` or something similar?

chrisbra commented on runtime/ftplugin/tigel.vim:
```diff
@@ -0,0 +1,13 @@
+" Vim filetype plugin file
+" Language:	TI DSP gel
+" Maintainer:	The Vim Project <https://github.com/vim/vim>
```
delete this line please

chrisbra commented on runtime/ftplugin/tisa.vim:
```diff
@@ -0,0 +1,13 @@
+" Vim filetype plugin file
+" Language:	TI linear assembly language
+" Maintainer:	The Vim Project <https://github.com/vim/vim>
```
and this one

chrisbra commented on runtime/ftplugin/tild.vim:
```diff
@@ -0,0 +1,13 @@
+" Vim filetype plugin file
+" Language:	TI DSP linker command file
+" Maintainer:	The Vim Project <https://github.com/vim/vim>
```
and this one

chrisbra commented on runtime/syntax/tigel.vim:
```diff
@@ -0,0 +1,14 @@
+" Vim syntax file
+" Language:	TI DSP gel
+" Maintainer:	The Vim Project <https://github.com/vim/vim>
```
and this

chrisbra commented on runtime/syntax/tild.vim:
```diff
@@ -0,0 +1,14 @@
+" Vim syntax file
+" Language:	TI DSP linker command file
+" Maintainer:	The Vim Project <https://github.com/vim/vim>
```
and this

chrisbra commented on runtime/syntax/tisa.vim:
```diff
@@ -0,0 +1,14 @@
+" Vim syntax file
+" Language:	TI linear assembly language
+" Maintainer:	The Vim Project <https://github.com/vim/vim>
```
and this

zeertzjq commented on src/testdir/test_filetype.vim:
```diff
@@ -659,7 +660,6 @@ def s:GetFilenameChecks(): dict<list<string>>
     samba: ['smb.conf'],
     sas: ['file.sas'],
     sass: ['file.sass'],
-    sather: ['file.sa'],
```
Does this line have to be removed? I think it will work fine.

zeertzjq commented on runtime/autoload/dist/ft.vim:
```diff
@@ -994,6 +994,14 @@ export def SQL()
 enddef
 
+export def FTsa()
+  if getline(1) =~# ';\*'
```
Is checking the first line enough?

chrisbra commented on runtime/autoload/dist/ft.vim:
```diff
@@ -994,6 +994,14 @@ export def SQL()
 enddef
 
+export def FTsa()
+  if getline(1) .. getline(2) .. getline(3) .. getline(4) =~# ';\*'
```
```suggestion
  if join(getline(1, 4), '') =~# ';\*'
```

chrisbra commented on src/testdir/test_filetype.vim:
```diff
@@ -659,7 +660,7 @@ def s:GetFilenameChecks(): dict<list<string>>
     sas: ['file.sas'],
     sass: ['file.sass'],
-    sather: ['file.sa'],
+    sather: ['file.sather'],
```
In `filetype.vim`, we are only detecting `*.sa` files as potential `sather filetype`. Did you mean to add a file pattern `*.sather` as well?

Freed-Wu committed:
filetype: TI assembly files are not recognized

Problem: TI assembly files are not recognized
Solution: recognize TI assembly files as tiasm

zeertzjq commented on src/testdir/test_filetype.vim:
```diff
@@ -2310,6 +2309,16 @@ func Test_cls_file()
+  call writefile([';* XXX-a.sa: XXX for TI C6000 DSP *;', '.no_mdep'], 'Xfile.sa')
+  split Xfile.sa
+  call assert_equal('linearasm', &filetype)
+
```
The `sather` fallback is not tested

zeertzjq commented on runtime/autoload/dist/ft.vim:
```diff
@@ -1003,6 +1003,14 @@ export def SQL()
 
+export def FTsa()
+  if join(getline(1, 4), '') =~# ';\*'
+    setf linearsam
```
```suggestion
    setf linearasm
```

zeertzjq commented on src/testdir/test_filetype.vim:
```diff
@@ -2309,6 +2308,21 @@ func Test_cls_file()
+  call writefile([';* XXX-a.sa: XXX for TI C6000 DSP *;', '.no_mdep'], 'Xfile.sa')
+  split Xfile.sa
+  call assert_equal('linearasm', &filetype)
+
```
```suggestion
  bwipe!

```

zeertzjq commented on runtime/syntax/tiasm.vim:
```diff
@@ -0,0 +1,102 @@
+hi def link tiasmType		Type
+hi def link tiasmFunction	Function
+
+let b:current_syntax = "lineartiasm"
```
Should this be changed to `tiasm` now?

Freed-Wu replied:
My typo. linear assemble is just a special type of TI asm, which use same syntax highlight.
